agent/taskresource/credentialspec/credentialspec_linux.go

```diff
@@ -0,0 +1,404 @@
+		seelog.Infof("grpc connection: %v", conn)
+		if err != nil {
+			seelog.Errorf("failed to connect with credentials fetcher daemon: %s", err)
+			return err
+		}
+		// make the grpc call to add kerberos lease api to create kerberos tickets for the gmsa account
+		response, err := credentialsfetcherclient.NewCredentialsFetcherClient(conn, time.Minute).AddKerberosLease(context.Background(), cs.credentialsFetcherRequest)
+
+		if err != nil {
+			seelog.Errorf("failed to create kerberos tickets associated service account, error: %s", err)
+			cs.setTerminalReason(err.Error())
+			return err
+		}
+
+		cs.leaseID = response.LeaseID
+		seelog.Infof("credentials fetcher response leaseID: %v", cs.leaseID)
+
+		//update the mapping of credspec ARN to the kerberos ticket location on the container instance
+		for _, kerberosTicketLocation := range response.KerberosTicketPaths {
+			for k, v := range cs.ServiceAccountInfoMap {
+				result := strings.Contains(strings.ToLower(kerberosTicketLocation), strings.ToLower(v.serviceAccountName))
+				if result {
+					cs.CredSpecMap[k] = kerberosTicketLocation
+					break
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (cs *CredentialSpecResource) handleCredentialspecFile(credentialSpec string, wg *sync.WaitGroup, errorEvents chan error) {
+	defer wg.Done()
+
+	credSpecSplit := strings.SplitAfterN(credentialSpec, "credentialspec:", 2)
+	if len(credSpecSplit) != 2 {
+		seelog.Errorf("Invalid credentialspec: %s", credentialSpec)
+		err := errors.New("invalid credentialspec file specification")
+		cs.setTerminalReason(err.Error())
+		errorEvents <- err
+		return
+	}
+	credSpecFile := credSpecSplit[1]
+
+	if !strings.HasPrefix(credSpecFile, "file://") {
+		err := errors.New("invalid credentialspec file specification")
+		cs.setTerminalReason(err.Error())
+		errorEvents <- err
+		return
+	}
+
+	fileName := strings.SplitAfterN(credSpecFile, "file://", 2)
+	data, err := ioutil.ReadFile(fileName[1])
+	if err != nil {
+		cs.setTerminalReason(err.Error())
+		errorEvents <- err
+		return
+	}
+
+	credSpecData := string(data)
+
+	cs.updateCredSpecMapping(credentialSpec, credSpecData)
+}
+
+func (cs *CredentialSpecResource) handleS3CredentialspecFile(originalCredentialSpec, credentialSpecS3ARN string, iamCredentials credentials.IAMRoleCredentials, wg *sync.WaitGroup, errorEvents chan error) {
+	defer wg.Done()
+	if iamCredentials == (credentials.IAMRoleCredentials{}) {
+		err := errors.New("credentialspec resource: unable to find execution role credentials")
+		cs.setTerminalReason(err.Error())
+		errorEvents <- err
+		return
+	}
+
+	_, err := arn.Parse(credentialSpecS3ARN)
+	if err != nil {
+		cs.setTerminalReason(err.Error())
+		errorEvents <- err
+		return
+	}
+
+	bucket, key, err := s3.ParseS3ARN(credentialSpecS3ARN)
+	if err != nil {
+		cs.setTerminalReason(err.Error())
+		errorEvents <- err
+		return
+	}
+
+	s3Client := cs.s3ClientCreator.NewS3Client(cs.region, iamCredentials)
```

Comment:
can you use the package's s3 / ssm clients directly (see envfile.go)?

https://github.com/aws/amazon-ecs-agent/blob/225bc3a556bd2d1759ab27b23f54e7e68086c9f0/agent/taskresource/envFiles/envfile.go#L121

https://github.com/aws/amazon-ecs-agent/blob/225bc3a556bd2d1759ab27b23f54e7e68086c9f0/agent/taskresource/envFiles/envfile.go#L73

Reply:
I was using it similar to that of envfile.go, in my case the resource is set in common file credentialspec.go